File: v4/rust/iprange-livedb/src/heap.rs
//! Exact accounting for caller-bounded retained heap.

use std::mem::size_of;

use crate::error::{Error, Result};

pub(crate) struct Heap {
    remaining: u64,
    used: u64,
}

impl Heap {
    pub(crate) const fn new(max_heap_bytes: u64) -> Self {
        Self {
            remaining: max_heap_bytes,
            used: 0,
        }
    }

    pub(crate) const fn used(&self) -> u64 {
        self.used
    }

    pub(crate) const fn remaining(&self) -> u64 {
        self.remaining
    }

    pub(crate) fn reserve_bytes(&mut self, bytes: usize, label: &'static str) -> Result<()> {
        let bytes = u64::try_from(bytes).map_err(|_| Error::BudgetExceeded(label))?;
        self.remaining = self
            .remaining
            .checked_sub(bytes)
            .ok_or(Error::BudgetExceeded(label))?;
        self.used = self
            .used
            .checked_add(bytes)
            .ok_or(Error::BudgetExceeded(label))?;
        Ok(())
    }

    pub(crate) fn vector<T>(&mut self, capacity: usize, label: &'static str) -> Result<Vec<T>> {
        self.charge::<T>(capacity, label)?;
        let mut values = Vec::new();
        values
            .try_reserve_exact(capacity)
            .map_err(|_| Error::BudgetExceeded(label))?;
        self.charge::<T>(values.capacity().saturating_sub(capacity), label)?;
        Ok(values)
    }

    pub(crate) fn filled<T: Clone>(
        &mut self,
        length: usize,
        value: T,
        label: &'static str,
    ) -> Result<Vec<T>> {
        let mut values = self.vector(length, label)?;
        values.resize(length, value);
        Ok(values)
    }

    fn charge<T>(&mut self, count: usize, label: &'static str) -> Result<()> {
        let bytes = count
            .checked_mul(size_of::<T>())
            .and_then(|bytes| u64::try_from(bytes).ok())
            .ok_or(Error::BudgetExceeded(label))?;
        self.remaining = self
            .remaining
            .checked_sub(bytes)
            .ok_or(Error::BudgetExceeded(label))?;
        self.used = self
            .used
            .checked_add(bytes)
            .ok_or(Error::BudgetExceeded(label))?;
        Ok(())
    }
}

File: v4/go/internal/writer/heap.go
// Bounded heap accounting for writer-side retained planning state (Rust
// heap.rs): one modeled budget per operation, charged with the exact Rust
// vector/filled byte counts under the draft MaxHeapBytes and the
// operation label, failing with the insufficient-resource class on
// overflow. Go allocations are separate and bounded by the same charges
// (the reader operationHeap precedent; Rust charges the exact Vec
// capacity, Go models the same logical capacity).

package writer

// heapBudget is one modeled per-operation heap (Rust Heap).
type heapBudget struct {
	remaining uint64
	used      uint64
}

// newHeapBudget starts one budget under the caller's byte limit (Rust
// Heap::new).
func newHeapBudget(maxHeapBytes uint64) *heapBudget {
	return &heapBudget{remaining: maxHeapBytes}
}

// usedBytes reports the charged bytes (Rust Heap::used).
func (h *heapBudget) usedBytes() uint64 { return h.used }

// remainingBytes reports the uncharged budget (Rust Heap::remaining).
func (h *heapBudget) remainingBytes() uint64 { return h.remaining }

// reserveBytes charges one exact byte reserve (Rust Heap::reserve_bytes).
func (h *heapBudget) reserveBytes(bytes uint64, label string) error {
	if bytes > h.remaining {
		return budgetExceeded(label)
	}
	h.remaining -= bytes
	h.used += bytes
	return nil
}

// vector charges one retained slice of count elements (Rust
// Heap::vector::<T>: count * size_of::<T>).
func (h *heapBudget) vector(count, elementBytes uint64, label string) error {
	return h.charge(count, elementBytes, label)
}

// filled charges one retained filled slice of length elements (Rust
// Heap::filled::<T>).
func (h *heapBudget) filled(length, elementBytes uint64, label string) error {
	return h.charge(length, elementBytes, label)
}

func (h *heapBudget) charge(count, elementBytes uint64, label string) error {
	if count != 0 && elementBytes > ^uint64(0)/count {
		return budgetExceeded(label)
	}
	return h.reserveBytes(count*elementBytes, label)
}
